### PR TITLE
improved markdown to html rendering using flexmark extensions

### PR DESCRIPTION
**Pull request by UBessle**
Implements the proposed changes from https://github.com/avisi-cloud/structurizr-site-generatr/issues/121

The configuration of the used extensions can be made using a property on workspace level
"generatr.markdown.flexmark.extensions" "Admonition"

This property has to be on workspace level due to the fact, that the rendered markdown text doesn't belong to any view. Instead it comes from workspace (and softwareSystem) level documentation files. So it seemed counter intuitive to place the corresponding property somewhere in the view definitions. Now it would be placed just above or below the definition where the workspace documents are.

As a future task it could be made possible, to have different settings for each softwareSystem.

Additionally, it is only possible to define which extensions to use without the possibility to use additional configuration options for some of the extensions. 
So no property 
"generatr.markdown.flexmark.Emoji.USE_IMAGE_TYPE" "UNICODE_ONLY"

### CODE REVIEW

**Comment by dirkgroot** [comment]
Hi @UBessle, thanks for your PR! I'm soon going to have a closer look. While quickly glancing over the changes, I noticed that you added a `bin` directory in the root. You also added this directory to `.gitignore`, so I assume you added this directory by accident. Could you please remove it, if that's the case?

**Comment by UBessle** [comment]
yes bin directory was a mistake. I've removed it from git.

**Comment by UBessle** [comment]
Hi @dirkgroot,
we wood love, to use all the flexmark extensions in our architecture documentations. On the other side we would like to stay with the official release of structurizr-site-generatr to get all those improvements. So it would be good to know, when you plan to get a closer look on this pull request.
Thanks, Uwe

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/PageViewModel.kt [hunk 1]
```diff
@@ -1,20 +1,28 @@
...
     val pageTitle: String by lazy {
         if (pageSubTitle.isNotBlank() && generatorContext.workspace.name.isNotBlank())
-            "$pageSubTitle | ${generatorContext.workspace.name}"
+        "$pageSubTitle | ${generatorContext.workspace.name}"
         else if (generatorContext.workspace.name.isNotBlank())
-            generatorContext.workspace.name
+        generatorContext.workspace.name
         else
-            pageSubTitle
+        pageSubTitle
```
Comment (dirkgroot) [inline src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/PageViewModel.kt:13]: Please make sure that this code is properly indented. This is easily done by using the auto formatter of your IDE of choice (again: I recommend using IntelliJ Community edition).

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/PageViewModel.kt [hunk 2]
```diff
@@ -1,20 +1,28 @@
...
     }
     val favicon by lazy { FaviconViewModel(generatorContext, this) }
     val headerBar by lazy { HeaderBarViewModel(this, generatorContext) }
     val menu by lazy { MenuViewModel(generatorContext, this) }
     val includeAutoReloading = generatorContext.serving
+    
+    val flexmarkConfig:FlexmarkConfig by lazy {
```
Comment (dirkgroot) [inline src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/PageViewModel.kt:20]:
```suggestion
    val flexmarkConfig by lazy { 
```

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/MarkdownExtension.kt [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+fun BODY.markdownAdmonitionScript(viewModel: PageViewModel) {
+    script(
+        type = ScriptType.textJavaScript,
+        src = "../" + "/admonition.js".asUrlToFile(viewModel.url)
+    ) { }
+}
+
+fun HEAD.katexStylesheet() {
+    // loading KaTeX as global on a webpage: https://katex.org/docs/browser.html#loading-as-global
+    // <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.6/dist/katex.min.css" integrity="sha384-mXD7x5S50Ko38scHSnD4egvoExgMPbrseZorkbE49evAfv9nNcbrXJ8LLNsDgh9d" crossorigin="anonymous">
```
Comment (dirkgroot) [inline src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/MarkdownExtension.kt:23]: Please remove this line. It's just a commented-out copy of line 26.

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/MarkdownExtension.kt [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    unsafe { 
+        raw("""
+            <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.6/dist/katex.min.css" integrity="sha384-mXD7x5S50Ko38scHSnD4egvoExgMPbrseZorkbE49evAfv9nNcbrXJ8LLNsDgh9d" crossorigin="anonymous">
+        """) 
+    }
+}
+
+fun HEAD.katexScript() {
+    // loading KaTeX as global on a webpage: https://katex.org/docs/browser.html#loading-as-global
+    // <script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.6/dist/katex.min.js" integrity="sha384-j/ZricySXBnNMJy9meJCtyXTKMhIJ42heyr7oAdxTDBy/CYA9hzpMo+YTNV5C+1X" crossorigin="anonymous"></script>
```
Comment (dirkgroot) [inline src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/MarkdownExtension.kt:33]: Please remove this line. It's just a commented-out copy of line 36.

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/MarkdownExtension.kt [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    // window.WebFontConfig = {
+    //     custom: {
+    //     families: ['KaTeX_AMS', 'KaTeX_Caligraphic:n4,n7', 'KaTeX_Fraktur:n4,n7',
+    //         'KaTeX_Main:n4,n7,i4,i7', 'KaTeX_Math:i4,i7', 'KaTeX_Script',
+    //         'KaTeX_SansSerif:n4,n7,i4', 'KaTeX_Size1', 'KaTeX_Size2', 'KaTeX_Size3',
+    //         'KaTeX_Size4', 'KaTeX_Typewriter'],
+    //     },
+    // };
+    // </script>
+    // <script defer src="https://cdn.jsdelivr.net/npm/webfontloader@1.6.28/webfontloader.js" integrity="sha256-4O4pS1SH31ZqrSO2A/2QJTVjTPqVe+jnYgOWUVr7EEc=" crossorigin="anonymous"></script>
```
Comment (dirkgroot) [inline src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/MarkdownExtension.kt:53]: Please remove this.

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/MarkdownExtension.kt [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            <script defer src="https://cdn.jsdelivr.net/npm/webfontloader@1.6.28/webfontloader.js" integrity="sha256-4O4pS1SH31ZqrSO2A/2QJTVjTPqVe+jnYgOWUVr7EEc=" crossorigin="anonymous"></script>
+        """)
+    }
+}
+
+fun BODY.mermaidScript() {
+    // Simple full example, how to include Mermaid: https://mermaid.js.org/config/usage.html#simple-full-example
+    // <script type="module">
+    //   import mermaid from 'https://cdn.jsdelivr.net/npm/mermaid@10/dist/mermaid.esm.min.mjs';
+    // </script>
```
Comment (dirkgroot) [inline src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/MarkdownExtension.kt:75]: Please remove this.

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/MarkdownExtension.kt [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    script(
+        type = "module"
+    ) { 
+        +"import mermaid from 'https://cdn.jsdelivr.net/npm/mermaid@10/dist/mermaid.esm.min.mjs';" 
+    }
```
Comment (dirkgroot) [inline src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/MarkdownExtension.kt:80]: The unary `+` operator on `script` is deprecated. Use `unsafe` instead.

```suggestion
    script(type = "module") {
        unsafe {
            raw("import mermaid from 'https://cdn.jsdelivr.net/npm/mermaid@10/dist/mermaid.esm.min.mjs';")
        }
    }
```

**Review by dirkgroot — CHANGES_REQUESTED** [review]
Thanks again for your contribution! My apologies that this took such a long time. I've added a bunch of comments to your code, most of which are about cleaning up the code.

Looking at the resulting possibilies, I really like this addition!

File: docs/example/workspace.dsl [hunk 8]
```diff
@@ -5,6 +5,14 @@
  * - "Big Bank plc - Internet Banking System" (https://structurizr.com/share/36141/)
 */
 workspace "Big Bank plc" "This is an example workspace to illustrate the key features of Structurizr, via the DSL, based around a fictional online banking system." {
+    properties {
+        // full list of available "generatr.markdown.flexmark.extensions" 
+        // "Abbreviation,Admonition,AnchorLink,Aside,Attributes,Autolink,Definition,Emoji,EnumeratedReference,Footnotes,GfmIssues,GfmStrikethroughSubscript,GfmTaskList,GfmUsers,GitLab,Ins,Macros,MediaTags,ResizableImage,Superscript,Tables,TableOfContents,SimulatedTableOfContents,Typographic,WikiLinks,XWikiMacro,YAMLFrontMatter,YouTubeLink"
```
[the remark this answers is not in the document]
Comment (UBessle) [reply docs/example/workspace.dsl]: done

File: docs/example/workspace.dsl [hunk 9]
```diff
@@ -5,6 +5,14 @@
...
+        // ATTENTION: 
+        // * "generatr.markdown.flexmark.extensions" values must be separated by comma
+        // * it's not possible to use "GitLab" and "ResizableImage" extensions together
+        "generatr.markdown.flexmark.extensions" "Abbreviation,Admonition,AnchorLink,Attributes,Autolink,Definition,Emoji,Footnotes,GfmTaskList,GitLab,MediaTags,Tables,TableOfContents,Typographic"
```
[the remark this answers is not in the document]
Comment (UBessle) [reply docs/example/workspace.dsl]: done

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/FlexmarkConfig.kt [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    "Typographic" to com.vladsch.flexmark.ext.typographic.TypographicExtension.create(),
+    "WikiLinks" to com.vladsch.flexmark.ext.wikilink.WikiLinkExtension.create(),
+    "XWikiMacro" to com.vladsch.flexmark.ext.xwiki.macros.MacroExtension.create(),
+    "YAMLFrontMatter" to com.vladsch.flexmark.ext.yaml.front.matter.YamlFrontMatterExtension.create(),
+    // "YouTrackConverter" to com.vladsch.flexmark.ext.youtrack.converter.YouTrackConverterExtension.create(),
+    "YouTubeLink" to com.vladsch.flexmark.ext.youtube.embedded.YouTubeLinkExtension.create(),
+)
+
+fun buildFlexmarkConfig(generatorContext: GeneratorContext): FlexmarkConfig {
+    var flexmarkExtensionString: String = "Tables"
```
[the remark this answers is not in the document]
Comment (UBessle) [reply src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/FlexmarkConfig.kt]: done

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/MarkdownExtension.kt (hunk 3 above)
Comment (UBessle) [reply src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/MarkdownExtension.kt]: done

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/MarkdownExtension.kt (hunk 4 above)
Comment (UBessle) [reply src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/MarkdownExtension.kt]: done

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/MarkdownExtension.kt (hunk 5 above)
Comment (UBessle) [reply src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/MarkdownExtension.kt]: done

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/MarkdownExtension.kt (hunk 6 above)
Comment (UBessle) [reply src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/MarkdownExtension.kt]: done

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/MarkdownExtension.kt (hunk 7 above)
Comment (UBessle) [reply src/main/kotlin/nl/avisi/structurizr/site/generatr/site/views/MarkdownExtension.kt]: done

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/FlexmarkConfig.kt [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+import nl.avisi.structurizr.site.generatr.site.GeneratorContext
+
+data class FlexmarkConfig(
+    val flexmarkExtensionsProperty: String,
+    val selectedExtensionMap: Map<String, Extension>,
+    val flexmarkOptions: DataHolder
+)
+
+val availableExtensionMap: MutableMap<String, Extension> = mutableMapOf<String, Extension>(
```
[the remark this answers is not in the document]
Comment (UBessle) [reply src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/FlexmarkConfig.kt]: done

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/FlexmarkConfig.kt
[the remark this answers is not in the document]
Comment (UBessle) [reply src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/FlexmarkConfig.kt:1]: done

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/FlexmarkConfig.kt [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+    var flexmarkExtensionString: String = "Tables"
+
+    if (generatorContext.workspace.properties.containsKey("generatr.markdown.flexmark.extensions")) {
+        flexmarkExtensionString = generatorContext.workspace.properties["generatr.markdown.flexmark.extensions"] as String
+    }
+
+    val flexmarkExtensionNames = flexmarkExtensionString.split(",")
+
+    // Create a Mutable List to store Extension objects
+    val selectedExtensionMap: MutableMap<String, Extension> = mutableMapOf<String, Extension>();
```
[the remark this answers is not in the document]
Comment (UBessle) [reply src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/FlexmarkConfig.kt]: done

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/FlexmarkConfig.kt [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+    
+    // Add each extension to the MutableDataSet
+    flexmarkExtensionNames.forEach { it ->
```
[the remark this answers is not in the document]
Comment (UBessle) [reply src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/FlexmarkConfig.kt]: done

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/FlexmarkConfig.kt [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+        val extensionName = it.trim()
+        if (availableExtensionMap.containsKey(extensionName)) {
+            selectedExtensionMap.put(extensionName, availableExtensionMap[extensionName] as Extension)
```
[the remark this answers is not in the document]
Comment (UBessle) [reply src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/FlexmarkConfig.kt]: done

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/MarkdownToHtml.kt [hunk 15]
```diff
@@ -18,10 +18,12 @@ import org.jsoup.Jsoup
 import org.jsoup.nodes.Element
 
 fun markdownToHtml(pageViewModel: PageViewModel, markdown: String, svgFactory: (key: String, url: String) -> String?): String {
-    val options = MutableDataSet()
-
-    options.set(Parser.EXTENSIONS, listOf(TablesExtension.create()))
-
+    // val options = MutableDataSet()
+    
+    // options.set(Parser.EXTENSIONS, listOf(TablesExtension.create()))
```
[the remark this answers is not in the document]
Comment (UBessle) [reply src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/MarkdownToHtml.kt:23]: done

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/MarkdownToHtml.kt
[the remark this answers is not in the document]
Comment (UBessle) [reply src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/MarkdownToHtml.kt:1]: done

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/MarkdownToHtml.kt [hunk 16]
```diff
@@ -18,10 +18,12 @@ import org.jsoup.Jsoup
...
+    val flexmarkConfig: FlexmarkConfig = pageViewModel.flexmarkConfig
```
[the remark this answers is not in the document]
Comment (UBessle) [reply src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/MarkdownToHtml.kt:24]: done

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/PageViewModel.kt [hunk 17]
```diff
@@ -1,20 +1,28 @@
 package nl.avisi.structurizr.site.generatr.site.model
 
 import nl.avisi.structurizr.site.generatr.site.GeneratorContext
+import nl.avisi.structurizr.site.generatr.site.model.FlexmarkConfig
```
[the remark this answers is not in the document]
Comment (UBessle) [reply src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/PageViewModel.kt:4]: done

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/PageViewModel.kt (hunk 1 above)
Comment (UBessle) [reply src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/PageViewModel.kt:13]: done

File: src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/PageViewModel.kt (hunk 2 above)
Comment (UBessle) [reply src/main/kotlin/nl/avisi/structurizr/site/generatr/site/model/PageViewModel.kt:20]: done

**Comment by UBessle** [comment]
Hi @dirkgroot ,
thanks for the very detailed feedback to my PR and sorry for the fact, that there was so much room for improvement. Unfortunately Kotlin was a totally new environment for me and so it is no big surprise that this shows somewhat in the quality of my code. So thanks again, that you did invest so much time in giving feedback, to improve it.
Cheers, Uwe

**Comment by UBessle** [comment]
I've tried to follow your suggestion to use Intellij for the work on the Kotlin Code. Unfortunately for some reasons all github projects reside in WSL2 (Windows Subsystem for Linux) on my Windows 10 workplace machine and Intellij does not work very well in combination with WSL2. It works, but it is awfull slow when accessing the source files. So the automatic indexing in IDEA runs for hours already and it hasn't reached 50% yet. So it remains a little bit tricky for me to get more productive in this setup ...

**Review by dirkgroot — APPROVED** [review]

**Pull request merged**